streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -731,15 +732,20 @@ private void addToTasksToClose(final Map<TaskId, CompletableFuture<StateUpdater.
                                    final Set<Task> tasksToCloseCleanFromStateUpdater,
                                    final Set<Task> tasksToCloseDirtyFromStateUpdater) {
         futures.entrySet().stream()
-            .map(entry -> waitForFuture(entry.getKey(), entry.getValue()))
-            .filter(Objects::nonNull)
-            .forEach(removedTaskResult -> {
-                if (removedTaskResult.exception().isPresent()) {
-                    tasksToCloseDirtyFromStateUpdater.add(removedTaskResult.task());
-                } else {
-                    tasksToCloseCleanFromStateUpdater.add(removedTaskResult.task());
-                }
-            });
+                .map(entry -> waitForFuture(entry.getKey(), entry.getValue()))
+                .filter(Objects::nonNull)
+                .forEach(removedTaskResult -> {
+                    if (removedTaskResult.exception().isPresent()) {
+                        final RuntimeException runtimeException = removedTaskResult.exception().get();
+                        if (runtimeException instanceof TaskCorruptedException) {
```

Comment:
In general, we prefer to be conservative and if in doubt close-dirty to ensure we never violate correctness. Thus, I am wondering if we should flip this logic, and explicitly "allow list" for what type of error we want to close clean?

This PR basically default to "close clean" and "deny list" a clean close by only closing dirty for `TaskCorruptedException`.

What type of exception did you see on your side, for which we know we can safely close clean? If these exception are scattered in the code base, we could also add `try-catch` on these places and re-throw a new (eg) `TaskFailedException` which we can catch here, and which would tell us that "close clean" is ok.

Thoughts?

Reply:
Indeed, we are flipping the condition that allows play conservative regarding data integrity, but from what I see, there are a few cases where the state store of an updating task might be inconsistent, and one of them is the `TaskCorruptedException`, which is basically an `InvalidOffsetException`, or when a task tries to initialize the state and the checkpoint file is not found. It is also possible to receive an `IllegalStateException`, but I think that's only possible in case of a bug in the code.

I agree that we should maintain a conservative mindset, particularly because restoring tasks will eventually turn into an active store, making integrity crucial. I noticed that `StreamsException`s are the most common exceptions that are being thrown in the state updater flow, and I think we can cleanly close the task when this exception is thrown, so that will cover timeouts and generic Kafka client exceptions. `InterruptedException` exceptions are also possible, and I recommend wrapping this exception in a StreamsException too. 